frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
+  }
+
+  //Default Node Selection strategy.
+  private SelectionResult selectNodesByJobPacking(ResourceDescriptor requestResource, int pendingTaskNumber) {
+
+    int requestNumber = pendingTaskNumber * am.getConfiguration().getLauncherConfig().getAmSearchNodeBufferFactor();
+    List<Node> candidateNodes = new ArrayList<Node>();
+    SelectionResult result = new SelectionResult();
+    for (String nodeName : filteredNodes) {
+      candidateNodes.add(allNodes.get(nodeName));
+    }
+    Collections.sort(candidateNodes);
+    for (int i = 0; i < requestNumber && i < candidateNodes.size(); i++) {
+      Node select = candidateNodes.get(i);
+      Long gpuAttribute = requestResource.getGpuAttribute();
+      if (gpuAttribute == 0) {
+        gpuAttribute = selectCandidateGpuAttribute(select, requestResource.getGpuNumber());
+      }
+      result.addSelection(select.getHost(), gpuAttribute, select.getAvailableResource().getPortRanges());
     }
+    return result;
+  }
 
-    if (selectionResult != null) {
-      LOGGER.logInfo(
-          "select: Found a SelectionResult satisfies the Request: SelectionResult: [%s]",
-          selectionResult);
-    } else {
-      LOGGER.logWarning(
-          "select: Cannot found a SelectionResult satisfies the Request: " +
-              "Check whether the Request can be relaxed to RM");
-      String notRelaxLogPrefix = "select: The Request cannot be relaxed to RM: Reason: ";
-
-      // Test NodeGpuType
-      if (requestNodeGpuType != null) {
-        throw new NotAvailableException(
-            String.format(notRelaxLogPrefix +
-                    "NodeGpuType is specified: Request NodeGpuType: [%s]",
-                requestNodeGpuType));
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String taskRoleName) throws NotAvailableException {
+
+    LOGGER.logInfo(
+        "select: TaskRole: [%s] Resource: [%s]", taskRoleName, requestResource);
+    String requestNodeLabel = am.getRequestManager().getTaskPlatParams().get(taskRoleName).getTaskNodeLabel();
+    String requestNodeGpuType = am.getRequestManager().getTaskPlatParams().get(taskRoleName).getTaskNodeGpuType();
+    int pendingTaskNumber = am.getStatusManager().getUnAssociatedTaskCount(taskRoleName);
+    List<ValueRange> reUsePorts = null;
+
+    // Prefer to use previous successfully allocated ports. if no successfully ports, try to re-use the "Requesting" ports.
+    if (am.getRequestManager().getTaskRoles().get(taskRoleName).getUseTheSamePorts()) {
+      reUsePorts = am.getStatusManager().getLiveAssociatedContainerPorts(taskRoleName);
+      if (ValueRangeUtils.getValueNumber(reUsePorts) <= 0 && requestedPortsCache.containsKey(taskRoleName)) {
+        reUsePorts = requestedPortsCache.get(taskRoleName);
+        // the cache only guide the next task to use previous requesting port.
+        requestedPortsCache.remove(taskRoleName);
       }
+    }
```

Comment:
>} [](start = 4, length = 1)

maybe you can add an else to log a warning: not implemented (for not use the same port)

Reply:
For not use the same Port, the code already implemented :).
If not use the same ports,   The behavior is "Random "allocated for each task. 

And there is also Log in line:206 to log the reused port information. 


---
In reply to: [173424356](https://github.com/Microsoft/pai/pull/318#discussion_r173424356) [](ancestors = 173424356)